### PR TITLE
Add fix LSTM

### PR DESCRIPTION
**Pull request by tienhxh**
Create a "get_lstm_dataloader" function.

### CODE REVIEW

File: models/regression/simple_lstm.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+class BasicLSTM(BaseModel):
+    """
+    Basic LSTM model.
+
+    :param input_dim: Input dimension.
+    :param hidden_dim: Hidden state dimension.
+    :param num_layers: Number of LSTM layers.
+    :param output_dim: Output dimension.
+    :param seq_length: Sequence length.
+    """
```
Comment (nguyenhongson1996) [inline models/regression/simple_lstm.py:42]: Move this block to inside the __init__ function.

File: utils/lstm_dataset.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            for substance in CHEMICAL_SUBSTANCE_COLUMNS
+        }
+
+    def __len__(self):
+        # Total number of sequences that can be generated
+        return len(self.samples) - self.seq_length + 1
+
+    def __getitem__(self, idx):
+        # Create a sequence from the dataset
+        sequence_samples = self.samples[idx: idx + self.seq_length]
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:31]: Here we will return overlap samples, correct? For example, we have data of 12 months and the sequence length is 6. Then the first sample will start from Jan to June, the second sample will start from Feb to July?

File: utils/lstm_dataset.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                    sample.chem_substance_concentration[substance] 
+                    if not math.isnan(sample.chem_substance_concentration[substance])
+                    else self.nan_default_replace_value[substance]
+                    for substance in CHEMICAL_SUBSTANCE_COLUMNS
+                ],
+                dtype=torch.float32
+            ) for sample in sequence_samples
+        ])
+
+        # The target for the sequence is the target value of the last sample
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:46]: Are we get the last sample or the label of the entire sequence?

File: utils/lstm_dataset.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+from utils.data_utils import DataSample, get_avg_value
+from utils.consts import CHEMICAL_SUBSTANCE_COLUMNS
+
+class ChemicalSequenceDataset(Dataset):
+    """
+    Dataset for LSTM: Returns sequences with a fixed length.
+    """
+
+    def __init__(self, samples: List[DataSample], seq_length: int):
+        # Filter out samples with NaN target values
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:14]:
```suggestion
        # Filter out all samples with NaN as target.
```

File: utils/lstm_dataset.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        samples = [sample for sample in samples if not math.isnan(sample.target_value)]
+        self.samples = samples
+        self.seq_length = seq_length  # Length of each sequence
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:17]:
```suggestion
        self.seq_length = seq_length  # Sequence length.
```

File: utils/lstm_dataset.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+        # Replace NaN values in chemical concentrations with average values
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:19]:
```suggestion
        # Replace NaN values in chemical concentrations with the corresponding average value.
```

File: utils/lstm_dataset.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        self.nan_default_replace_value = {
+            substance: get_avg_value(substance=substance, samples=self.samples)
+            for substance in CHEMICAL_SUBSTANCE_COLUMNS
+        }
+
+    def __len__(self):
+        # Total number of sequences that can be generated
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:26]:
```suggestion
        # Total number of sequences that can be generated.
```

File: utils/lstm_dataset.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        return len(self.samples) - self.seq_length + 1
+
+    def __getitem__(self, idx):
+        # Create a sequence from the dataset
+        sequence_samples = self.samples[idx: idx + self.seq_length]
+
+        # Extract features and targets for the entire sequence
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:33]:
```suggestion
        # Extract features and targets for the entire sequence.
```

File: utils/lstm_dataset.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            dtype=torch.float32
+        )  # Shape: (seq_length,)
+        return features, targets
+
+def get_lstm_dataloader(data_by_location: Dict[str, List[DataSample]], 
+                        batch_size: int = 4, seq_length: int = 5, 
+                        shuffle: bool = False) -> DataLoader:
+    """
+    Create a DataLoader for LSTM.
+    :param data_by_location: Dictionary where keys are locations and values are sample lists.
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:58]:
```suggestion
    :param data_by_location: Dictionary with key being the location and value being the samples.
```

File: utils/lstm_dataset.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    :param batch_size: Batch size.
+    :param seq_length: Length of each sequence.
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:60]:
```suggestion
    :param seq_length: Sequence length.
```

File: utils/lstm_dataset.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    :param shuffle: Whether to shuffle the data.
+    :return: Dataloader.
+    """
+    # Combine all samples from different locations
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:64]:
```suggestion
    # Combine the samples from different locations.
```

File: utils/lstm_dataset.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    all_samples = [
+        sample for location in sorted(list(data_by_location)) 
+        for sample in data_by_location[location]
+    ]
+    
+    # Create a dataset with a fixed sequence length
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:70]:
```suggestion
    # Create a dataset with a fixed sequence length.
```

File: utils/lstm_dataset.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    dataset = ChemicalSequenceDataset(all_samples, seq_length)
+
+    # Create a DataLoader
```
Comment (nguyenhongson1996) [inline utils/lstm_dataset.py:73]:
```suggestion
    # Create a DataLoader.
```

**Review by nguyenhongson1996 — CHANGES_REQUESTED** [review]

**Review by nguyenhongson1996 — APPROVED** [review]
LGTM

File: models/regression/simple_lstm.py [hunk 14]
```diff
@@ -66,6 +74,26 @@ def calculate_detailed_report(self, predictions: List[torch.Tensor], ground_trut
...
+        """
+        W = self.model.fc.weight.data  # (output_dim, hidden_dim)
+        c = self.model.fc.bias.data    # (output_dim,)
+
+        y = y.unsqueeze(0) 
+        c = c.unsqueeze(0) # (batch_size, output_dim)
+
+        # Tính nghịch đảo của W (pseudo-inverse)
+        W_inv = torch.pinverse(W)  # (hidden_dim, output_dim)
+        # Tính toán x từ y: x = (y - c) * W^-1
```
Comment (nguyenhongson1996) [inline models/regression/simple_lstm.py:91]: Is this formular correct? Should it be X=W^-1(y-c)?

File: models/regression/simple_lstm.py [hunk 15]
```diff
@@ -96,7 +124,10 @@ def fit(self, train_loader: DataLoader, val_loader: Optional[DataLoader], epochs
 
             for batch_x, batch_y in train_loader:
                 optimizer.zero_grad()
-                y_pred = self(batch_x)  # Predict the output for the entire sequence.
+                y_prev = batch_y[0, 0]
+                h0_new = self.reconstruct_hidden_state(y_prev)
+                # Truyền h0 mới vào LSTM
+                y_pred = self(batch_x, h0=h0_new)
```
Comment (nguyenhongson1996) [inline models/regression/simple_lstm.py:130]: c0 is zero here then this condition is line61 is always True. Then h0 means nothing in this code.

**Review by nguyenhongson1996 — CHANGES_REQUESTED** [review]
H0 means nothing. Please re-check.

**Review by nguyenhongson1996 — APPROVED** [review]
LGTM
